### PR TITLE
Fix the scourge of game add-on build failures

### PR DESCRIPTION
**Pull request by garbear**
## Description

As can be seen at https://jenkins.kodi.tv/view/Game%20Binary%20Addons/, not many add-ons are green. This is because about 3/4 of the time on Windows, calling `prepare-env.bat` fails during git clean because gpg-agent.exe is daemonized in the background. An example is [here](https://jenkins.kodi.tv/blue/organizations/jenkins/kodi-game%2Fgame.libretro.bsnes/detail/115.0.0.13-Nexus/1/pipeline/282).

To fix this and allow `prepare-env.bat` to run successfully, I've added a line to kill the daemon.

## Motivation and context

Fix game binary add-ons failing 3/4 of the time.

## How has this been tested?

Started by building 2048:

```
PS C:\Users\thema\Documents\kodi-master\tools\buildsteps\windows\x64> .\prepare-env.bat
PS C:\Users\thema\Documents\kodi-master\tools\buildsteps\windows\x64> .\make-addons.bat package game.libretro.2048
```

Then daemonized gpg-agent.exe:

```
PS C:\Users\thema\Documents\kodi-master\cmake\addons\build\mingw\src\mingw\usr\bin> .\gpg-agent.exe --daemon
```

Verified it couldn't be removed:

```
PS C:\Users\thema\Documents\kodi-master\cmake\addons\build\mingw\src\mingw\usr\bin> rm .\gpg-agent.exe
rm : Cannot remove item C:\Users\thema\Documents\kodi-master\cmake\addons\build\mingw\src\mingw\usr\bin\gpg-agent.exe:
Access to the path 'C:\Users\thema\Documents\kodi-master\cmake\addons\build\mingw\src\mingw\usr\bin\gpg-agent.exe' is
denied.
At line:1 char:1
+ rm .\gpg-agent.exe
+ ~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : PermissionDenied: (C:\Users\thema\...n\gpg-agent.exe:FileInfo) [Remove-Item], Unauthoriz
   edAccessException
    + FullyQualifiedErrorId : RemoveFileSystemItemUnAuthorizedAccess,Microsoft.PowerShell.Commands.RemoveItemCommand
```

Ran `prepare-env.bat`, viewed the failure:

```
PS C:\Users\thema\Documents\kodi-master\tools\buildsteps\windows\x64> .\prepare-env.bat
Workspace is C:\Users\thema\Documents\kodi-master
running git clean -xffd -e "project/BuildDependencies/downloads" -e "project/BuildDependencies/downloads2" -e "project/BuildDependencies/mingwlibs" -e "project/BuildDependencies/msys64" -e "project/BuildDependencies/tools"
Unlink of file 'cmake/addons/build/mingw/src/mingw/usr/bin/gpg-agent.exe' failed. Should I try again? (y/n)
```

Applied the patch, ran `prepare-env.bat` again, verified it completed successfully:

```
PS C:\Users\thema\Documents\kodi-master\tools\buildsteps\windows\x64> .\prepare-env.bat
Workspace is C:\Users\thema\Documents\kodi-master
running git clean -xffd -e "project/BuildDependencies/downloads" -e "project/BuildDependencies/downloads2" -e "project/BuildDependencies/mingwlibs" -e "project/BuildDependencies/msys64" -e "project/BuildDependencies/tools"
Removing cmake/addons/build/
Removing cmake/addons/output/
Removing project/BuildDependencies/win32/
Removing project/BuildDependencies/x64/
```

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

### CODE REVIEW

File: tools/buildsteps/windows/prepare-env.bat [hunk 1]
```diff
@@ -13,11 +13,14 @@ IF EXIST %WORKSPACE%\project\Win32BuildSetup\BUILD_WIN32 rmdir %WORKSPACE%\proje
...
 
+rem daemonized gpg-agent blocks mingw from being cleaned with git
+TASKKILL /IM "gpg-agent.exe" /F >nul 2>&1
+
 rem we assume git in path as this is a requirement
 rem git clean the untracked files and directories
 rem but keep the downloaded dependencies
 rem also keeps MSYS2 installation
-SET GIT_CLEAN_CMD=git clean -xffd -e "project/BuildDependencies/downloads" -e "project/BuildDependencies/downloads2" -e "project/BuildDependencies/mingwlibs" -e "project/BuildDependencies/msys64" -e "project/BuildDependencies/tools"
+SET GIT_CLEAN_CMD=git clean -xffd -e "project/BuildDependencies/downloads" -e "project/BuildDependencies/downloads2" -e "project/BuildDependencies/mingwlibs" -e "project/BuildDependencies/msys64" -e "project/BuildDependencies/tools" -e "cmake/addons/build/download/msys2-base-x86_64-20231026.tar.xz"
```
Comment (lrusak) [inline tools/buildsteps/windows/prepare-env.bat:23]: Does this need to use a regex so it doesn't have to be updated whenever msys2 changes version?

Comment (garbear) [reply tools/buildsteps/windows/prepare-env.bat:23]: It looks like git clean doesn't accept regexes. There's not much cost in not changing this if the msys2 version changes, as it only takes ~20s to download. We only update msys2 every few years so I don't think it's a big issue.

Comment (fuzzard) [reply tools/buildsteps/windows/prepare-env.bat:23]: why would you want to exclude a clean of this file explicitly anyway?

Comment (fuzzard) [reply tools/buildsteps/windows/prepare-env.bat:23]: * wildcard seems to work fine locally.

```
c:\Users\brent\source\repos\xbmc>notepad msys-base-x86_64-20231026.tar.xz

c:\Users\brent\source\repos\xbmc>notepad msys2-base-x86_64-20231026.tar.xz

c:\Users\brent\source\repos\xbmc>dir *.tar.xz
 Volume in drive C has no label.
 Volume Serial Number is C07A-15D6

 Directory of c:\Users\brent\source\repos\xbmc

19/01/2024  05:06 PM                 0 msys-base-x86_64-20231026.tar.xz
19/01/2024  05:03 PM                 0 msys2-base-x86_64-20231026.tar.xz
               2 File(s)              0 bytes
               0 Dir(s)  140,702,515,200 bytes free

c:\Users\brent\source\repos\xbmc>git clean -xffd -e msys2-base-x86_64-*.tar.xz
Removing msys-base-x86_64-20231026.tar.xz

c:\Users\brent\source\repos\xbmc>dir *.tar.xz
 Volume in drive C has no label.
 Volume Serial Number is C07A-15D6

 Directory of c:\Users\brent\source\repos\xbmc

19/01/2024  05:03 PM                 0 msys2-base-x86_64-20231026.tar.xz
               1 File(s)              0 bytes
               0 Dir(s)  140,702,457,856 bytes free

c:\Users\brent\source\repos\xbmc>git --version
git version 2.35.1.windows.2
```

Comment (garbear) [reply tools/buildsteps/windows/prepare-env.bat]: Wildcard works fine.

Comment (garbear) [reply tools/buildsteps/windows/prepare-env.bat]: The file is needed by 160 game add-ons. I think it makes sense to cache a 57MB download on every run.

Comment (garbear) [reply tools/buildsteps/windows/prepare-env.bat]: I tried to cache the pacman packages too. I added the path `cmake\addons\build\mingw\src\mingw\var\cache\pacman\pkg` to excludes, but the state would get funky when it re-ran. It still downloads another 50MB of packages, it'd be nice to be able to cache those too.

**Comment by garbear** [comment]
Looks like the Omega branch works: https://jenkins.kodi.tv/blue/organizations/jenkins/kodi-game%2Fgame.libretro.2048/detail/master/145/pipeline/73/

Now we need a Nexus backport: https://github.com/xbmc/xbmc/pull/24562

**Pull request merged**